File: android/app/src/main/java/cz/drabek/chuvicka/baby/BabyServer.kt
package cz.drabek.chuvicka.baby

import cz.drabek.chuvicka.Log
import cz.drabek.chuvicka.proto.RtpPacketizer
import cz.drabek.chuvicka.proto.interleaved
import java.io.BufferedInputStream
import java.io.IOException
import java.io.InputStream
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.atomic.AtomicLong
import kotlin.random.Random

/**
 * A small RTSP server on the phone at the baby. It speaks the same RTSP as go2rtc and as the
 * iOS app's BabyServer: RTP interleaved over TCP, H.264 video and G.711 A-law sound, and
 * `GET /<code>/frame.jpeg` for a photo. So an iPhone parent and an Android parent both read it.
 *
 * The pairing code is the first part of the path. A wrong code gets 401. After 10 wrong codes
 * in a minute, all requests fail for a minute.
 */
class BabyServer(
    private val code: String,
    private val hasVideo: Boolean,
    private val onClients: (all: Int, video: Int) -> Unit,
    private val onNeedKeyframe: () -> Unit,
    private val onFrameRequest: () -> ByteArray?,
) {
    private var server: ServerSocket? = null
    private val sessions = ConcurrentHashMap.newKeySet<Session>()
    @Volatile private var sps: ByteArray? = null
    @Volatile private var pps: ByteArray? = null
    private val wrongCodes = ArrayList<Long>()
    @Volatile private var running = false

    /** It opens a port that the system chooses, and returns it for the mDNS registration. */
    fun start(): Int {
        // A fixed port, so the address that a parent keeps for the time away stays valid.
        val s = try { ServerSocket(BABY_PORT) } catch (e: IOException) { ServerSocket(0) }
        server = s
        running = true
        Thread({
            while (running) {
                val socket = try { s.accept() } catch (e: IOException) { break }
                socket.tcpNoDelay = true
                val session = Session(socket)
                sessions.add(session)
                session.start()
            }
        }, "baby-accept").apply { isDaemon = true; start() }
        Log.add("baby server on port ${s.localPort}")
        return s.localPort
    }

    private val port get() = server?.localPort ?: BABY_PORT

    /** This phone's IPv4 addresses, the Tailscale one first. A parent keeps them for the time away. */
    private fun addresses(): String = cz.drabek.chuvicka.Net.ipv4().joinToString(", ") { "$it:$port" }

    fun stop() {
        running = false
        try { server?.close() } catch (_: IOException) {}
        sessions.forEach { it.close() }
        sessions.clear()
        report()
    }

    // MARK: The media

    fun sendVideo(nals: List<ByteArray>, timestamp: Long, keyframe: Boolean, sps: ByteArray?, pps: ByteArray?) {
        if (sps != null && pps != null) { this.sps = sps; this.pps = pps }
        val s = this.sps
        val p = this.pps
        // The parameter sets in the stream too: a parent that joins late needs no new DESCRIBE.
        val unit = if (keyframe && s != null && p != null) listOf(s, p) + nals else nals
        for (session in sessions) if (session.playing) session.sendVideo(unit, timestamp, keyframe)
    }

    /** 20 ms of A-law sound (160 bytes at 8 kHz). */
    fun sendAudio(aLaw: ByteArray, timestamp: Long) {
        for (session in sessions) if (session.playing) session.sendAudio(aLaw, timestamp)
    }

    private fun report() {
        val playing = sessions.filter { it.playing }
        onClients(playing.size, playing.count { it.takesVideo })
    }

    @Synchronized
    private fun authorized(uri: String): Boolean {
        val now = System.currentTimeMillis()
        wrongCodes.removeAll { now - it > 60_000 }
        if (wrongCodes.size >= 10) return false
        if (pathCode(uri) == code) return true
        wrongCodes.add(now)
        Log.add("baby server: a phone with a wrong code")
        return false
    }

    /** "rtsp://host/482913?audio/trackID=1" or "/482913/frame.jpeg" to "482913". */
    private fun pathCode(uri: String): String {
        val afterHost = if ("://" in uri) uri.substringAfter("://").let { r -> r.indexOf('/').let { if (it < 0) "" else r.substring(it) } } else uri
        return afterHost.split("/").firstOrNull { it.isNotEmpty() }?.substringBefore("?") ?: ""
    }

    private fun sdp(audioOnly: Boolean): String {
        val b = StringBuilder("v=0\r\no=- 0 0 IN IP4 0.0.0.0\r\ns=Chuvicka\r\nt=0 0\r\n")
        if (hasVideo && !audioOnly) {
            b.append("m=video 0 RTP/AVP 96\r\na=rtpmap:96 H264/90000\r\n")
            var fmtp = "packetization-mode=1"
            val s = sps
            val p = pps
            if (s != null && p != null) {
                val enc = { x: ByteArray -> android.util.Base64.encodeToString(x, android.util.Base64.NO_WRAP) }
                fmtp += ";sprop-parameter-sets=${enc(s)},${enc(p)}"
            }
            b.append("a=fmtp:96 $fmtp\r\na=control:trackID=0\r\n")
        }
        b.append("m=audio 0 RTP/AVP 8\r\na=rtpmap:8 PCMA/8000\r\na=control:trackID=1\r\n")
        return b.toString()
    }

    /** One parent: a reader thread for the requests, and a writer thread with a bounded queue. */
    private inner class Session(private val socket: Socket) {
        @Volatile var playing = false
        @Volatile var videoChannel = -1
        @Volatile var audioChannel = -1
        val takesVideo get() = videoChannel >= 0
        private val sessionId = Random.nextInt(100_000, Int.MAX_VALUE).toString()
        private val video = RtpPacketizer(96)
        private val audio = RtpPacketizer(8)
        private val queue = LinkedBlockingQueue<ByteArray>()
        private val queued = AtomicLong(0)
        @Volatile private var waitForKeyframe = true
        @Volatile private var closed = false

        fun start() {
            Thread({ read() }, "baby-read").apply { isDaemon = true; start() }
            Thread({ write() }, "baby-write").apply { isDaemon = true; start() }
        }

        fun close() {
            if (closed) return
            closed = true
            val was = playing
            playing = false
            try { socket.close() } catch (_: IOException) {}
            queue.offer(ByteArray(0))              // It wakes the writer.
            sessions.remove(this)
            if (was) { Log.add("parent left"); report() }
        }

        // MARK: Sending

        @Synchronized
        fun sendVideo(nals: List<ByteArray>, timestamp: Long, keyframe: Boolean) {
            val channel = videoChannel
            if (channel < 0) return
            if (waitForKeyframe) {
                if (!keyframe) return
                waitForKeyframe = false
            }
            // More than about 1 s of picture waits: skip the picture until the next keyframe.
            if (queued.get() > 256 * 1024) {
                waitForKeyframe = true
                onNeedKeyframe()
                return
            }
            for (p in video.h264(nals, timestamp)) enqueue(interleaved(p, channel))
        }

        @Synchronized
        fun sendAudio(payload: ByteArray, timestamp: Long) {
            val channel = audioChannel
            if (channel < 0 || queued.get() > 512 * 1024) return
            enqueue(interleaved(audio.packet(payload, 0, payload.size, timestamp, false), channel))
        }

        private fun enqueue(bytes: ByteArray) {
            queued.addAndGet(bytes.size.toLong())
            queue.offer(bytes)
        }

        private fun write() {
            try {
                val out = socket.getOutputStream()
                while (!closed) {
                    val b = queue.take()
                    if (b.isEmpty()) continue
                    out.write(b)
                    queued.addAndGet(-b.size.toLong())
                    if (queue.isEmpty()) out.flush()
                }
            } catch (_: Exception) {
                close()
            }
        }

        // MARK: The requests

        private fun read() {
            try {
                val input = BufferedInputStream(socket.getInputStream())
                while (!closed) {
                    val first = input.read()
                    if (first < 0) break
                    if (first == 0x24) {                      // RTCP from the parent: skip it.
                        input.read(); val len = (input.read() shl 8) or input.read()
                        skip(input, len)
                        continue
                    }
                    val head = readHead(input, first) ?: break
                    val lines = head.split("\r\n")
                    val request = lines[0].split(" ")
                    val headers = HashMap<String, String>()
                    for (line in lines.drop(1)) {
                        val c = line.indexOf(':')
                        if (c > 0) headers[line.substring(0, c).lowercase()] = line.substring(c + 1).trim()
                    }
                    headers["content-length"]?.toIntOrNull()?.let { skip(input, it) }
                    if (request.size >= 3) handle(request[0], request[1], request[2], headers)
                }
            } catch (_: Exception) {
            }
            close()
        }

        /** InputStream.skipNBytes needs Android 13. This works on Android 8. */
        private fun skip(input: InputStream, n: Int) {
            var left = n
            while (left > 0) {
                if (input.read() < 0) return
                left--
            }
        }

        private fun readHead(input: InputStream, first: Int): String? {
            val b = StringBuilder().append(first.toChar())
            while (!b.endsWith("\r\n\r\n")) {
                val c = input.read()
                if (c < 0 || b.length > 16 * 1024) return null
                b.append(c.toChar())
            }
            return b.toString().trimEnd()
        }

        private fun handle(method: String, uri: String, version: String, headers: Map<String, String>) {
            val cseq = headers["cseq"] ?: "0"
            if (version.startsWith("HTTP")) {
                val jpeg = if (method == "GET" && uri.endsWith("/frame.jpeg") && authorized(uri)) onFrameRequest() else null
                if (jpeg != null) http("200 OK", "image/jpeg", jpeg) else http("401 Unauthorized", "text/plain", "no".toByteArray())
                close()
                return
            }
            if (method != "OPTIONS" && method != "GET_PARAMETER" && !authorized(uri)) {
                reply(401, "Unauthorized", cseq); return
            }
            when (method) {
                "OPTIONS" -> reply(200, "OK", cseq, mapOf("Public" to "OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN, GET_PARAMETER"))
                "GET_PARAMETER" -> reply(200, "OK", cseq, mapOf("Session" to sessionId))
                "DESCRIBE" -> reply(200, "OK", cseq,
                    mapOf("Content-Type" to "application/sdp", "X-Chuvicka-Addresses" to addresses()), sdp("?audio" in uri))
                "SETUP" -> {
                    val isVideo = "trackID=0" in uri
                    val transport = headers["transport"] ?: ""
                    val channel = Regex("interleaved=(\\d+)").find(transport)?.groupValues?.get(1)?.toIntOrNull()
                        ?: if (isVideo) 0 else 2
                    if (isVideo) videoChannel = channel else audioChannel = channel
                    reply(200, "OK", cseq, mapOf(
                        "Transport" to "RTP/AVP/TCP;unicast;interleaved=$channel-${channel + 1}",
                        "Session" to "$sessionId;timeout=60"))
                }
                "PLAY" -> {
                    reply(200, "OK", cseq, mapOf("Session" to sessionId, "Range" to "npt=0.000-"))
                    if (!playing) {
                        playing = true
                        waitForKeyframe = true
                        Log.add("parent connected")
                        report()
                        onNeedKeyframe()
                    }
                }
                "TEARDOWN" -> { reply(200, "OK", cseq, mapOf("Session" to sessionId)); close() }
                else -> reply(405, "Method Not Allowed", cseq)
            }
        }

        private fun reply(status: Int, reason: String, cseq: String, headers: Map<String, String> = emptyMap(), body: String = "") {
            val b = StringBuilder("RTSP/1.0 $status $reason\r\nCSeq: $cseq\r\nServer: Chuvicka\r\n")
            for ((k, v) in headers) b.append("$k: $v\r\n")
            val bytes = body.toByteArray(Charsets.UTF_8)
            if (bytes.isNotEmpty()) b.append("Content-Length: ${bytes.size}\r\n")
            b.append("\r\n")
            enqueue(b.toString().toByteArray(Charsets.UTF_8) + bytes)
        }

        private fun http(status: String, type: String, body: ByteArray) {
            val head = "HTTP/1.1 $status\r\nContent-Type: $type\r\nContent-Length: ${body.size}\r\nConnection: close\r\n\r\n"
            try {
                socket.getOutputStream().apply { write(head.toByteArray()); write(body); flush() }
            } catch (_: IOException) {}
        }
    }
}

/** The fixed port of the phone at the baby. The iOS app uses the same. */
const val BABY_PORT = 8555

File: Nursery/Baby/BabyServer.swift
import Foundation
import Network

/// The Bonjour type of the iPhone at the baby. It must also be in NSBonjourServices.
enum BabyService {
    static let type = "_chuvicka._tcp"
}

/// A small RTSP server on the iPhone at the baby. It speaks the same RTSP as go2rtc,
/// so the parent app reads it with the same client: RTP interleaved over TCP,
/// H.264 video and G.711 A-law sound. It also answers `GET /<code>/frame.jpeg` for a photo.
///
/// The phones find each other with Bonjour, on the home Wi-Fi or directly (peer-to-peer Wi-Fi).
/// Nothing goes to the internet. The pairing code is the path, so a phone without the code
/// gets 401 and no sound.
///
/// All the state is on `queue`.
final class BabyServer: @unchecked Sendable {
    let queue = DispatchQueue(label: "nursery.baby.server", qos: .userInteractive)

    /// The count of parents that receive the stream now. It runs on the main queue.
    var onClients: ((Int) -> Void)?
    /// A parent started to play. The encoder must make a keyframe now.
    var onNeedKeyframe: (() -> Void)?
    /// A parent asked for a photo. The callback gets a JPEG, or nil.
    var onFrameRequest: ((@escaping @Sendable (Data?) -> Void) -> Void)?

    private let code: String
    private let hasVideo: Bool
    private var listener: NWListener?
    private var sessions: [ObjectIdentifier: Session] = [:]
    private var sps: [UInt8]?
    private var pps: [UInt8]?

    init(code: String, video: Bool) {
        self.code = code
        self.hasVideo = video
    }

    // MARK: Start and stop

    func start(name: String) throws {
        let tcp = NWProtocolTCP.Options()
        tcp.noDelay = true
        tcp.enableKeepalive = true
        tcp.keepaliveIdle = 5
        let params = NWParameters(tls: nil, tcp: tcp)
        params.includePeerToPeer = true         // Also with no router: the phones connect directly.
        params.serviceClass = .interactiveVideo
        let listener = try NWListener(using: params)
        listener.service = NWListener.Service(name: name, type: BabyService.type)
        listener.newConnectionHandler = { [weak self] conn in self?.accept(conn) }
        listener.stateUpdateHandler = { state in
            if case .failed(let error) = state { Log.shared.add("baby server failed: \(error)") }
        }
        listener.start(queue: queue)
        self.listener = listener
        Log.shared.add("baby server started as \(name)")
    }

    func stop() {
        queue.async {
            self.listener?.cancel()
            self.listener = nil
            for s in self.sessions.values { s.close() }
            self.sessions.removeAll()
            self.reportClients()
        }
    }

    // MARK: The media, from the capture

    /// One H.264 access unit, with no start codes. The SPS and the PPS come with each keyframe.
    func sendVideo(nals: [[UInt8]], timestamp: UInt32, keyframe: Bool, sps: [UInt8]?, pps: [UInt8]?) {
        queue.async {
            if let sps, let pps { self.sps = sps; self.pps = pps }
            var unit = nals
            // The parameter sets in the stream too: a parent that joins late needs no new DESCRIBE.
            if keyframe, let s = self.sps, let p = self.pps { unit = [s, p] + nals }
            for s in self.sessions.values where s.playing { s.sendVideo(unit, timestamp: timestamp, keyframe: keyframe) }
        }
    }

    /// 20 ms of A-law sound (160 bytes at 8 kHz).
    func sendAudio(_ aLaw: [UInt8], timestamp: UInt32) {
        queue.async {
            for s in self.sessions.values where s.playing { s.sendAudio(aLaw, timestamp: timestamp) }
        }
    }

    // MARK: The connections

    private func accept(_ conn: NWConnection) {
        let session = Session(connection: conn, server: self)
        sessions[ObjectIdentifier(session)] = session
        conn.stateUpdateHandler = { [weak self, weak session] state in
            guard let self, let session else { return }
            switch state {
            case .ready: session.receive()
            case .failed, .cancelled: self.remove(session)
            default: break
            }
        }
        conn.start(queue: queue)
    }

    fileprivate func remove(_ session: Session) {
        guard sessions.removeValue(forKey: ObjectIdentifier(session)) != nil else { return }
        session.close()
        if session.playing { Log.shared.add("parent left") }
        reportClients()
    }

    fileprivate func reportClients() {
        let n = sessions.values.filter(\.playing).count
        DispatchQueue.main.async { self.onClients?(n) }
    }

    fileprivate func started(_ session: Session) {
        Log.shared.add("parent connected")
        reportClients()
        onNeedKeyframe?()
    }

    fileprivate func requestKeyframe() { onNeedKeyframe?() }

    // MARK: The requests

    fileprivate func pathIsValid(_ uri: String) -> Bool {
        // "rtsp://host/482913?audio/trackID=1" or "/482913/frame.jpeg": the code is the first path part.
        let afterHost: Substring
        if let range = uri.range(of: "://") {
            let rest = uri[range.upperBound...]
            afterHost = rest.firstIndex(of: "/").map { rest[$0...] } ?? ""
        } else {
            afterHost = uri[...]
        }
        let first = afterHost.split(separator: "/").first.map { $0.split(separator: "?").first ?? "" } ?? ""
        return first == code
    }

    fileprivate func sdp(audioOnly: Bool) -> String {
        var s = "v=0\r\no=- 0 0 IN IP4 0.0.0.0\r\ns=Chuvicka\r\nt=0 0\r\n"
        if hasVideo && !audioOnly {
            s += "m=video 0 RTP/AVP 96\r\na=rtpmap:96 H264/90000\r\n"
            var fmtp = "packetization-mode=1"
            if let sps, let pps {
                fmtp += ";sprop-parameter-sets=\(Data(sps).base64EncodedString()),\(Data(pps).base64EncodedString())"
            }
            s += "a=fmtp:96 \(fmtp)\r\na=control:trackID=0\r\n"
        }
        s += "m=audio 0 RTP/AVP 8\r\na=rtpmap:8 PCMA/8000\r\na=control:trackID=1\r\n"
        return s
    }

    fileprivate func frame(_ done: @escaping @Sendable (Data?) -> Void) {
        guard hasVideo, let onFrameRequest else { done(nil); return }
        onFrameRequest(done)
    }
}

/// One parent. It reads the requests, and sends RTP to the channels that SETUP chose.
private final class Session {
    let connection: NWConnection
    weak var server: BabyServer?
    private(set) var playing = false
    private var buffer: [UInt8] = []
    private var sessionID = String(UInt32.random(in: 100_000...UInt32.max))
    private var videoChannel: UInt8?
    private var audioChannel: UInt8?
    private var video = RTPPacketizer(payloadType: 96)
    private var audio = RTPPacketizer(payloadType: 8)
    /// The bytes given to the connection that it did not send yet. A slow Wi-Fi fills it.
    private var inFlight = 0
    private var waitForKeyframe = true
    private var closed = false

    init(connection: NWConnection, server: BabyServer) {
        self.connection = connection
        self.server = server
    }

    func close() {
        guard !closed else { return }
        closed = true
        playing = false
        connection.cancel()
    }

    func receive() {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 64 * 1024) { [weak self] data, _, complete, error in
            guard let self, !self.closed else { return }
            if let data { self.buffer.append(contentsOf: data); self.parse() }
            if error != nil || complete { self.server?.remove(self); return }
            self.receive()
        }
    }

    // MARK: Sending

    func sendVideo(_ nals: [[UInt8]], timestamp: UInt32, keyframe: Bool) {
        guard let channel = videoChannel else { return }
        if waitForKeyframe {
            guard keyframe else { return }
            waitForKeyframe = false
        }
        // Too much is waiting: skip the picture until the next keyframe. The sound goes on.
        if inFlight > 768 * 1024 {
            waitForKeyframe = true
            server?.requestKeyframe()
            return
        }
        var bytes: [UInt8] = []
        for p in video.h264(nals, timestamp: timestamp) { bytes += interleaved(p, channel: channel) }
        write(bytes)
    }

    func sendAudio(_ payload: [UInt8], timestamp: UInt32) {
        guard let channel = audioChannel, inFlight < 2 * 1024 * 1024 else { return }
        write(interleaved(audio.packet(payload[...], timestamp: timestamp, marker: false), channel: channel))
    }

    private func write(_ bytes: [UInt8]) {
        inFlight += bytes.count
        let n = bytes.count
        connection.send(content: Data(bytes), completion: .contentProcessed { [weak self] _ in
            self?.inFlight -= n        // The completion runs on the server queue.
        })
    }

    // MARK: The requests

    private func parse() {
        while !buffer.isEmpty {
            if buffer[0] == 0x24 {                         // Interleaved data from the parent (RTCP): skip it.
                guard buffer.count >= 4 else { return }
                let length = Int(buffer[2]) << 8 | Int(buffer[3])
                guard buffer.count >= 4 + length else { return }
                buffer.removeFirst(4 + length)
                continue
            }
            guard let end = headerEnd() else {
                if buffer.count > 16 * 1024 { server?.remove(self) }
                return
            }
            let head = String(decoding: buffer[0..<end], as: UTF8.self)
            var lines = head.components(separatedBy: "\r\n")
            let requestLine = lines.removeFirst().split(separator: " ")
            var headers: [String: String] = [:]
            for line in lines {
                guard let colon = line.firstIndex(of: ":") else { continue }
                headers[line[..<colon].lowercased()] = line[line.index(after: colon)...].trimmingCharacters(in: .whitespaces)
            }
            let bodyLength = Int(headers["content-length"] ?? "0") ?? 0
            guard buffer.count >= end + 4 + bodyLength else { return }
            buffer.removeFirst(end + 4 + bodyLength)
            guard requestLine.count >= 3 else { continue }
            handle(method: String(requestLine[0]), uri: String(requestLine[1]),
                   version: String(requestLine[2]), headers: headers)
        }
    }

    private func headerEnd() -> Int? {
        guard buffer.count >= 4 else { return nil }
        for i in 0...(buffer.count - 4) where buffer[i] == 13 && buffer[i + 1] == 10 && buffer[i + 2] == 13 && buffer[i + 3] == 10 {
            return i
        }
        return nil
    }

    private func handle(method: String, uri: String, version: String, headers: [String: String]) {
        guard let server else { return }
        let cseq = headers["cseq"] ?? "0"

        if version.hasPrefix("HTTP") {
            guard method == "GET", server.pathIsValid(uri), uri.hasSuffix("/frame.jpeg") else {
                httpReply("401 Unauthorized", type: "text/plain", body: Data("wrong code".utf8))
                return
            }
            server.frame { [weak self] jpeg in
                server.queue.async {
                    guard let self else { return }
                    if let jpeg { self.httpReply("200 OK", type: "image/jpeg", body: jpeg) }
                    else { self.httpReply("503 Service Unavailable", type: "text/plain", body: Data("no picture".utf8)) }
                }
            }
            return
        }

        if method != "OPTIONS" && method != "GET_PARAMETER" && !server.pathIsValid(uri) {
            reply(401, "Unauthorized", cseq: cseq)
            Log.shared.add("baby server: a phone with a wrong code")
            return
        }
        switch method {
        case "OPTIONS":
            reply(200, "OK", cseq: cseq, ["Public": "OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN, GET_PARAMETER"])
        case "GET_PARAMETER":
            reply(200, "OK", cseq: cseq, ["Session": sessionID])
        case "DESCRIBE":
            let sdp = server.sdp(audioOnly: uri.contains("?audio"))
            reply(200, "OK", cseq: cseq, ["Content-Type": "application/sdp"], body: sdp)
        case "SETUP":
            let transport = headers["transport"] ?? ""
            var channel: UInt8 = uri.contains("trackID=0") ? 0 : 2
            if let range = transport.range(of: "interleaved=") {
                let digits = transport[range.upperBound...].prefix { $0.isNumber }
                channel = UInt8(digits) ?? channel
            }
            if uri.contains("trackID=0") { videoChannel = channel } else { audioChannel = channel }
            reply(200, "OK", cseq: cseq, [
                "Transport": "RTP/AVP/TCP;unicast;interleaved=\(channel)-\(channel + 1)",
                "Session": "\(sessionID);timeout=60",
            ])
        case "PLAY":
            reply(200, "OK", cseq: cseq, ["Session": sessionID, "Range": "npt=0.000-"])
            if !playing {
                playing = true
                waitForKeyframe = true
                server.started(self)
            }
        case "TEARDOWN":
            reply(200, "OK", cseq: cseq, ["Session": sessionID])
            server.remove(self)
        default:
            reply(405, "Method Not Allowed", cseq: cseq)
        }
    }

    private func reply(_ status: Int, _ reason: String, cseq: String, _ headers: [String: String] = [:], body: String = "") {
        var text = "RTSP/1.0 \(status) \(reason)\r\nCSeq: \(cseq)\r\nServer: Chuvicka\r\n"
        for (k, v) in headers { text += "\(k): \(v)\r\n" }
        let bodyData = Array(body.utf8)
        if !bodyData.isEmpty { text += "Content-Length: \(bodyData.count)\r\n" }
        text += "\r\n"
        write(Array(text.utf8) + bodyData)
    }

    private func httpReply(_ status: String, type: String, body: Data) {
        let head = "HTTP/1.1 \(status)\r\nContent-Type: \(type)\r\nContent-Length: \(body.count)\r\nConnection: close\r\n\r\n"
        connection.send(content: Data(head.utf8) + body, completion: .contentProcessed { [weak self] _ in
            guard let self else { return }
            self.server?.remove(self)
        })
    }
}
